### PR TITLE
[feat] Support modelopt_mixed for Turing and Ampere via Marlin

### PR DESCRIPTION
**Pull request by ir1ka**
Leverages Marlin kernels to enable `modelopt_mixed` quantization support, extending compatibility to NVIDIA Turing and Ampere architectures.

Due to limitations in Marlin, tensor dimensions must be aligned; however, the output dimensions of certain layers in the `nvidia/NVIDIA-Nemotron-3-Super-120B-A12B-NVFP4` model were misaligned. Consequently, zero-padding was applied.
As a precautionary measure, zero-padding was restricted exclusively to layers utilizing FP8 per-tensor quantization.

@jinzhen-lin Could you please take a look at this PR when you have a moment?

#38776

PR #45295 has already implemented padding for marlin n/k, so only `modelopt_mixed` support for turing and ampere needs to be enabled.

## Purpose

Support model `nvidia/NVIDIA-Nemotron-3-Super-120B-A12B-NVFP4` for Turing or Ampere.

## Test Plan

Test model `nvidia/NVIDIA-Nemotron-3-Super-120B-A12B-NVFP4`.
Validated implementation using `nvidia/NVIDIA-Nemotron-3-Super-120B-A12B-NVFP4`, ensuring functional correctness and performance stability on 4x RTX3090.

Command: `vllm serve nvidia/NVIDIA-Nemotron-3-Super-120B-A12B-NVFP4 --served-model-name Nemotron-3-Super --tensor-parallel-size 4 --enable-expert-parallel --trust-remote-code --enable-auto-tool-choice --tool-call-parser qwen3_coder --reasoning-parser nemotron_v3 --kv-cache-memory-bytes 3G --max-model-len auto --async-scheduling --enable-prefix-caching --enable-chunked-prefill --max-num-seqs 4`

## Test Result

The `vllm serve` command launched successfully, and performance tests appear normal.
```log
vllm-nemotron-3-super-1  | (APIServer pid=1) INFO 04-04 13:15:32 [loggers.py:259] Engine 000: Avg prompt throughput: 2.9 tokens/s, Avg generation throughput: 46.3 tokens/s, Running: 1 reqs, Waiting: 0 reqs, GPU KV cache usage: 1.6%, Prefix cache hit rate: 0.0%
vllm-nemotron-3-super-1  | (APIServer pid=1) INFO:     127.0.0.1:35860 - "POST /v1/chat/completions HTTP/1.1" 200 OK
vllm-nemotron-3-super-1  | (APIServer pid=1) INFO 04-04 13:15:42 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 18.1 tokens/s, Running: 0 reqs, Waiting: 0 reqs, GPU KV cache usage: 0.0%, Prefix cache hit rate: 0.0%
vllm-nemotron-3-super-1  | (APIServer pid=1) INFO 04-04 13:15:52 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 0.0 tokens/s, Running: 0 reqs, Waiting: 0 reqs, GPU KV cache usage: 0.0%, Prefix cache hit rate: 0.0%
vllm-nemotron-3-super-1  | (APIServer pid=1) INFO:     127.0.0.1:46700 - "POST /v1/chat/completions HTTP/1.1" 200 OK
vllm-nemotron-3-super-1  | (APIServer pid=1) INFO 04-04 13:18:42 [loggers.py:259] Engine 000: Avg prompt throughput: 2.2 tokens/s, Avg generation throughput: 11.8 tokens/s, Running: 0 reqs, Waiting: 0 reqs, GPU KV cache usage: 0.0%, Prefix cache hit rate: 0.0%
vllm-nemotron-3-super-1  | (APIServer pid=1) INFO 04-04 13:18:52 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 0.0 tokens/s, Running: 0 reqs, Waiting: 0 reqs, GPU KV cache usage: 0.0%, Prefix cache hit rate: 0.0%
```

Previously, the model `nvidia/NVIDIA-Nemotron-3-Super-120B-A12B-NVFP4` failed to launch because the `modelopt_mixed` constraint restricted deployment to SM89 or newer architectures only.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/model_executor/layers/quantization/utils/marlin_utils_fp8.py [hunk 1]
```diff
@@ -107,10 +129,30 @@ def prepare_fp8_layer_for_marlin(
     part_size_k = layer.input_size_per_partition
     weight_block_size = getattr(layer, "weight_block_size", None)
 
+    n_size, _ = marlin_pad(part_size_n, part_size_k)
+    pad_n = n_size - part_size_n
```
[the remark this answers is not in the document]
Comment (ir1ka) [reply vllm/model_executor/layers/quantization/utils/marlin_utils_fp8.py]: The `K` dimension padding involves complex transformations, and since no models currently requiring `K` dimension padding have been identified, we will temporarily omit `K` dimension padding in this PR.

File: csrc/quantization/marlin/awq_marlin_repack.cu [hunk 2]
```diff
@@ -221,11 +221,13 @@ __global__ void awq_marlin_repack_kernel(
 torch::Tensor awq_marlin_repack(torch::Tensor& b_q_weight, int64_t size_k,
                                 int64_t size_n, int64_t num_bits,
                                 bool is_a_8bit) {
+  const int target_tile_n_size = marlin::tile_n_size / (is_a_8bit ? 2 : 1);
```
Comment (robertgshaw2-redhat) [inline csrc/quantization/marlin/awq_marlin_repack.cu]: please explain this change

Comment (ir1ka) [reply csrc/quantization/marlin/awq_marlin_repack.cu]: I noticed that within the `gptq_marlin_repack_kernel`, the logic appears to be concerned solely with the expression `target_tile_n_size = marlin::tile_n_size / (is_a_8bit ? 2 : 1)`; consequently, I attempted to modify the preceding checks. This may have been a misunderstanding on my part, and I appreciate you pointing that out. I will revert the changes.
Given that the `lm-eval` tests have exhibited significant anomalies regarding accuracy and perplexity scores, I am converting this PR to a draft until the debugging process is complete.

**Comment by ir1ka** [comment]
It appears that while the Nemotron model defines numerous sub-layers (nn.Module), these layers are not included in the output of `model.named_modules()`; consequently, the `process_weights_after_loading` method for some of these layers is not being invoked.

**Comment by ir1ka** [comment]
Based on the mainline version (`0.19.2rc1.dev66+gb47840019`), the acc/PPL degradation issue has been resolved.

The test methods and results are as follows:

vLLM start command:
```bash
vllm serve nvidia/NVIDIA-Nemotron-3-Super-120B-A12B-NVFP4 --served-model-name Nemotron-3-Super --tensor-parallel-size 4 --enable-expert-parallel --trust-remote-code --enable-auto-tool-choice --tool-call-parser qwen3_coder --reasoning-parser nemotron_v3 --kv-cache-memory-bytes 2g --max-model-len auto --async-scheduling --enable-prefix-caching --enable-chunked-prefill --max-num-seqs 4 --mamba_ssm_cache_dtype float32 --seed 42
```

Test result of `lm-eval`:
```bash
# command
lm_eval --model local-completions  --model_args base_url=http://127.0.0.1:8000/v1/completions,model=Nemotron-3-Super,tokenizer=nvidia/NVIDIA-Nemotron-3-Super-120B-A12B-NVFP4 --trust_remote_code --tasks lambada_openai --batch_size auto --seed 42

# results
local-completions ({'base_url': 'http://127.0.0.1:8000/v1/completions', 'model': 'Nemotron-3-Super', 'tokenizer': 'nvidia/NVIDIA-Nemotron-3-Super-120B-A12B-NVFP4'}), gen_kwargs: ({}), limit: None, num_fewshot: None, batch_size: auto
|    Tasks     |Version|Filter|n-shot|  Metric  |   |Value |   |Stderr|
|--------------|------:|------|-----:|----------|---|-----:|---|-----:|
|lambada_openai|      1|none  |     0|acc       |↑  |0.7623|±  |0.0059|
|              |       |none  |     0|perplexity|↓  |3.0024|±  |0.0603|
```

File: csrc/quantization/marlin/awq_marlin_repack.cu (hunk 2 above)
Comment (ir1ka) [reply csrc/quantization/marlin/awq_marlin_repack.cu]: I have not yet fully understood the purpose of the code here, so I have reverted it.

**Comment by ir1ka** [comment]
@robertgshaw2-redhat Hello, this PR is ready for review.
Thanks.

File: vllm/model_executor/layers/quantization/utils/marlin_utils_fp8.py [hunk 3]
```diff
@@ -53,11 +66,13 @@ def apply_fp8_marlin_linear(
     # For GPUs that lack FP8 hardware support, we can leverage the
     # Marlin kernel for fast weight-only FP8 quantization
 
+    n_size, _ = marlin_pad(size_n, size_k)
```
Comment (jinzhen-lin) [inline vllm/model_executor/layers/quantization/utils/marlin_utils_fp8.py]: I think `padded_size_n` is a better variable name.

File: vllm/model_executor/layers/quantization/utils/marlin_utils_fp8.py
Comment (jinzhen-lin) [inline vllm/model_executor/layers/quantization/utils/marlin_utils_fp8.py:1]: Do gptq / awq / nvfp4 / mxfp4 have similar issues as well?

File: vllm/model_executor/layers/quantization/modelopt.py [hunk 4]
```diff
@@ -466,6 +466,7 @@ def create_weights(
         layer.logical_widths = output_partition_sizes
         layer.input_size_per_partition = input_size_per_partition
         layer.output_size_per_partition = output_size_per_partition
+        layer.orig_dtype = params_dtype
```
Comment (jinzhen-lin) [inline vllm/model_executor/layers/quantization/modelopt.py:476]: https://github.com/vllm-project/vllm/blob/9f771b3ab92d26a7d91a8255572c5d8d2b3ad601/vllm/model_executor/layers/linear.py#L264-L268

We have ensured that `layer.params_dtype` is set in `LinearBase`. So we can always use `layer.params_dtype` instead of defining another variable.

File: vllm/model_executor/layers/quantization/utils/marlin_utils_fp8.py
Comment (ir1ka) [reply vllm/model_executor/layers/quantization/utils/marlin_utils_fp8.py:1]: It appears that AWQ is not affected by this issue; the model I tested was `cyankiwi/NVIDIA-Nemotron-3-Super-120B-A12B-AWQ-4bit`.
The problem may stem from differences in weight sharding logic or the specific quantization method used (`modelopt_mixed`).

I am currently unable to test the GPTQ, NVFP4, and MXFP4 methods, as I do not have access to corresponding models for them.
NVFP4 should theoretically be affected as well; however, I did not encounter this issue during my testing, so I will not be making any modifications regarding it for the time being. The model `nvidia/NVIDIA-Nemotron-3-Super-120B-A12B-NVFP4` is a mixed-precision model utilizing NVFP4, FP8, and BF16; while this theoretically included a test of NVFP4, it did not trigger the issue.

This problem originates from `gptq_marlin_repack`, which requires that the N and K dimensions be aligned. This operator is invoked by the following files: `vllm/model_executor/layers/quantization/utils/marlin_utils_fp4.py`, `vllm/model_executor/layers/quantization/utils/marlin_utils_fp8.py`, and `vllm/model_executor/kernels/linear/mixed_precision/marlin.py`. However, `marlin_gemm` does not appear to perform any checks for this alignment; I am currently unsure whether the `marlin_gemm` operator itself imposes this same requirement.

**Comment by ir1ka** [comment]
The scope of this PR might be too broad; should I split it into two separate PRs?
1. One specifically for adding k-dim padding support for Marlin FP8;
2. The other to enable `modelopt_mixed` support for Ampere, building upon the first PR.
